### PR TITLE
[WIP] Expanded vector module -- spherical coordinate system and laplacian

### PR DESCRIPTION
**Pull request by jbbskinny**
# Summary

I am working on expanding the vector module (via [Upabjojr PR](https://github.com/sympy/sympy/pull/9937)).  I added a test suite for the new laplacian function and removed the coordinate system dependence for differential functions in _vector/functions.py_.  The tests run and pass with one exception.  The problem there is that if I use the gradient function on a number it is not able to find the system.  I believe that the when creating the Vector instances we should have a system attribute (same for VectorAdd and VectorMul).  This will make allow for some refactoring in _vector/functions.py_.
# ToDo:
- [ ] Possibly have a class per coordinate system.  DRY might be lost in library code, but it would look a lot cleaner in user code, each class could have its own differential method, etc.
- [ ] Have the differential coefficients initialize themselves when making new CoordSystem3D() instance.
- [ ] Carry a copy of CoodSystem3D() instance when creating Vector() instances (Add and Mul included).
- [ ] Add more coordinate systems.

### CODE REVIEW

File: doc/src/modules/vector/api/classes.rst [hunk 1]
```diff
@@ -2,13 +2,13 @@
 Essential Classes in sympy.vector (doctrings)
 =============================================
 
-CoordSysCartesian
-=================
+CoordSystem3D
```
Comment (moorepants) [inline doc/src/modules/vector/api/classes.rst]: Keep in mind that if you change the public API we have to go through deprecation cycles. If this module is in 0.7.6, that will be the case. See https://github.com/sympy/sympy/wiki/Deprecating-policy

Comment (moorepants) [inline doc/src/modules/vector/api/classes.rst]: If we really need to change the name, this would be better in a separate PR, because it makes it difficult to find what your PR title suggests you are changing.

Comment (jbbskinny) [inline doc/src/modules/vector/api/classes.rst]: Honestly, I had not thought about it.  I do feel that the DRY principle could be blown if we have different classes for each coordinate system.

Comment (jbbskinny) [inline doc/src/modules/vector/api/classes.rst]: It would definitely look better in user code!

Comment (Upabjojr) [inline doc/src/modules/vector/api/classes.rst]: In my PR it was just a renaming, you could avoid breaking compatibility by just adding:
```
def CoordSysCartesian(arg1, *args, **kwargs):
    return CoordSys3D(arg1, new_arg, *args, **kwargs)
```

Comment (Upabjojr) [inline doc/src/modules/vector/api/classes.rst]: By the way... we should keep in mind that this module could potentially be expanded in _N_ dimensions in the future. It's good to engineer the class-structure in a proper way.

File: sympy/vector/functions.py [hunk 2]
```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
...
 
     """
+    # Check input
+    if isinstance(scalar, Zero) or scalar == 0:
+        return Vector.zero
+
+    if not issubclass(type(scalar), Expr) or scalar.is_Vector:
+        raise TypeError("Your input should be a scalar field.")
+
+    for atom in scalar.atoms():
```
Comment (Upabjojr) [inline sympy/vector/functions.py]: `for atom in scalar.atoms(BaseScalar):`
so you don't need the _if_ after this

File: sympy/vector/functions.py [hunk 3]
```diff
@@ -196,24 +216,68 @@ def gradient(scalar, coord_sys):
...
+        if isinstance(atom, BaseScalar):
+            system = atom.system
+            break
```
Comment (Upabjojr) [inline sympy/vector/functions.py]: I would add a check that all atoms belong to the same coordinate system, and raise an exception if not.

Or better yet, make it belong to systems which have the same differential operators, e.g. a coord system and a translated instance of it.

Comment (jbbskinny) [inline sympy/vector/functions.py]: I considered that already, and it broke the doctests.

**Comment by jbbskinny** [comment]
For vectors I was going to add something like this to Vector():
```python
    @property
    def system(self):
        system = set()
        for ind, vec in self.components.items():
            system.add(ind.system)
        return list(system)
```

But when building a scalar field there is no core Scalar class that would allow this hack.

File: sympy/vector/functions.py (hunk 3 above)
Comment (Upabjojr) [inline sympy/vector/functions.py]: If you use PyCharm, you can debug the doctests quite easily. Better find out why, because it doesn't add up to me.

Comment (jbbskinny) [inline sympy/vector/functions.py]: The original author engineered it that way.  I believe the vectors defined should be well-defined vectors, that is, having only one coordinate system.

Comment (Upabjojr) [inline sympy/vector/functions.py]: Your code has a problem: it will use the _gradient_ operator of the first coordinate system it finds. If the gradient functions of the two scalars are different, it will return a wrong result.

@srjoglekar246 what do you think?

File: doc/src/modules/vector/api/classes.rst (hunk 1 above)
Comment (srjoglekar246) [inline doc/src/modules/vector/api/classes.rst]: +1. We need to think about the class structure for the coordinate systems classes a little better. What I had in mind when I designed it was to have a CoordSys base class, which defines the skeleton features of any coordinate system in sympy.vector. If needed, as @Upabjojr suggests, we could even support N-dimensions. Every type of coordinate system would then specialize the base class' properties and define their own features.

File: sympy/vector/functions.py (hunk 3 above)
Comment (srjoglekar246) [inline sympy/vector/functions.py]: I agree. If we really want to have every vector be defined in only _one_ coordinate system, then its better to remove the coordinate system parameter from this function as this PR does it. Otherwise, it would be better to leave it in.

Comment (Upabjojr) [inline sympy/vector/functions.py]: One simple example, suppose you want to calculate the gradient (in 3D) of `x*r`, you should calculate either:

* the gradient of `x*sqrt(x**2 + y**2 + z**2)`
* the gradient of `r**2*sin(theta)*cos(phi)`

The results are expressed in different coordinates, but should agree.

A simpler example, calculate the gradient of `x*xshift` where `xshift = x + delta` is a scalar field of a shifted coordinate system. The result is `(2*x + delta)*v_x` or `(2*xshift - delta)*v_xshift`.

This PR is already too large to handle all these cases. Maybe it's better not to address this point for now.

File: doc/src/modules/vector/api/classes.rst (hunk 1 above)
Comment (Upabjojr) [inline doc/src/modules/vector/api/classes.rst]: I'm not so favorable to using subclasses. This means that every coordinate system would require the implementation of its own class. I prefer to have a single class for all coordinate systems, the specific coordinate systems being its objects.

File: sympy/vector/functions.py (hunk 3 above)
Comment (jbbskinny) [inline sympy/vector/functions.py]: I hear you on that.  What we should focus on now is the original commits.  The removal of system parameter from *functions.py* should be seperate.  As @srjoglekar246 mentioned in a private message was that he designed CoordSysCartesian() to be a subclass of a general CoordSystem(R=3) class.  So maybe we should just keep that and have it inherited by the CoordSystem(R=3, sys=Cartesian) superclass if the user wants the system to be cartesian.  

What we need to decide on:
1. Do we want to restrict the CoordSystem() to Euclidian space? -- Depreciation process could be avoided in doing so. 
2. Have the superclass CoordSystem(R=3, sys=spherical) build the differential class within.

In future PR's:
1. Should we restrict vectors and scalar fields to one coordinate system at a time? -- Does Mathematica allow ```gradient(x*r)```?
2. Removal of the system requirement from *functions.py*
3. Vector functions and handling therein for *functions.py* -- ```type(f(C.x*C.i))``` gives ```f```
4. ...

Please note I'm just spitballing ideas here.  Feel free to add or tell me if it's a terrible idea.

File: doc/src/modules/vector/api/classes.rst (hunk 1 above)
Comment (jbbskinny) [inline doc/src/modules/vector/api/classes.rst]: The issue with the one class is that every time a new coordinate system is initialized is the we would have to define the differential_class which could become cumbersome.  If we do implement subclasses the Lambda() part could be avoided for known systems.  If a general CurvilinearSys() subclass is implemented (in a future PR) the user would need to define certain parameters for the coordinate system they want.

Comment (srjoglekar246) [inline doc/src/modules/vector/api/classes.rst]: It depends on how much we would need to pack into a single class. If theres too much of case-based(Spherical/Euclidean/Cylindrical) processing to be done, the methods in a single class could bloat up pretty easily, making it difficult to follow. Plus, having too many parameters to a single initializer isn't better as compared to having different properly-labeled classes. Ofcourse this is looking into the future, just my 2 cents.

Comment (Upabjojr) [inline doc/src/modules/vector/api/classes.rst]: > The issue with the one class is that every time a new coordinate system is initialized is that the user would have to define the differential_class which could become cumbersome.

Users could use shortcuts or factory functions, or special abbreviations to be passed to the constructor. In SymPy, constructor parameters have to be intended as the instance defining properties.

For example, you may add a parser transforming `differential_class="spherical"` to the proper _Lambda_ expression.

> the methods in a single class could bloat up pretty easily, making it difficult to follow.

The differential operators need not be defined separately for every coordinate system, they are generally defined in terms of the metric of the coordinate system.

When you subclass, you are supposed to define new methods and/or fields in the subclass. What are the new methods you are defining for spherical coordinates wrt Cartesian? None, they are the same. Of course, if someone doesn't know how to write code for differential operators general to all coordinate systems, he may end up rewriting the code in specific methods for every coordinate system. I started my PR in order to avoid this possibility to happen in the future, and only use a single class, as that's the correct way to proceed and avoid messing up SymPy.

The only point in favor of creating a subclass is the special case of 3 dimensions, where you can define the _curl_ which is not generalizable to _N_ dimensions (or at least, not easily/uniquely generalizable). But even in this case, it would be easier to use the same class and just raise an exception if _curl_ is invoked in _N != 3_ dimensions.

File: sympy/vector/functions.py (hunk 3 above)
Comment (Upabjojr) [inline sympy/vector/functions.py]: > 1. Should we keep CoordSysCartesian() as a subclass of an add superclass CoordSystem()?

No, this would be a violation of the DRY principle, and a useless repetition of code.

> 2. Do we want to restrict the CoordSystem() to Euclidian space?

Yes, for non-Euclidean geometries such as Riemannian and pseudo-Riemannian geometries there's _sympy.diffgeom_, which is specifically enginnered for that.

Typical operators of vector calculus have to be defined in a completely different way in pure differential geometry.

> 3. Have the superclass CoordSystem(R=3, sys=spherical) build the differential class within.

Use classes to correspond to mathematical objects. I would strongly avoid using two levels of classes to represent the same mathematical concept. You should create subclasses if you are restricting to a specific subset of coordinate systems which would require more methods to add to the class. If not so, DO NOT CREATE SUBCLASSES.

Comment (jbbskinny) [inline sympy/vector/functions.py]: > Use classes to correspond to mathematical objects.
> 
Nicely put.  Well, what do we need to do to get your original PR through?

What I’m thinking:
Have to depreciate CoordSysCartesian().
Add parser for differential_class for spherical and cylindrical coordinate systems.

Let me know what you think and I’ll work to get the commits in.  I will fix and add the updated functions.py file in a new PR once yours is through.

Future:
Need to add handling for spherical coordinates when integrating.  Not sure where this should take place just yet.
Needs to properly handle laplacian(1/r, spherical)

Justin

Comment (jbbskinny) [inline sympy/vector/functions.py]: > Use classes to correspond to mathematical objects.

Nicely put.  Well, what do we need to do to get your original PR through?

What I’m thinking:
1. Have to depreciate CoordSysCartesian().
2. Add parser for differential_class for spherical and cylindrical coordinate systems.
3. scalar_map() will need to be fixed

Example (for item 3):
In [1]: C = CoordSystem3D('C')
In [2]: spherical = CoordSystem3D('spherical', parent=C...) #Fixes origin at C's origin
...
In   [5]: spherical.scalar_map(C)
Out[5]: {spherical_phi: C_z, spherical_r: C_x, spherical_theta: C_y}

Let me know what you think and I’ll work to get the commits in.  I will fix and add the updated *functions.py* file in a new PR once yours is through.

Future:
* Need to add handling for spherical coordinates when integrating.  Not sure where this should take place just yet.
* Needs to properly handle laplacian(1/r, spherical)

Comment (jbbskinny) [inline sympy/vector/functions.py]: Instead of defining a different_class we should define a scalar map to the general system via a set a functions:

x = f(q1,q2,q3)
y = g(q1,q2,q3)
z = h(q1,q2,q3)

And, have the class work from there.  All differential coefficients can be determined by setting the scalar map from the start.  It also leaves room for generality, letting the user choose there own coordinate system and easy transition to N-dimensions.  Certainly, some scalar maps will be included (spherical and cylindrical).  Also, an inversion check will be added (determinant of the inverse jacobian is not zero), and hopefully, they will know to use smooth transformations.

I think this is the best route.

Comment (Upabjojr) [inline sympy/vector/functions.py]: > Well, what do we need to do to get your original PR through?

* Create tests for everything, spherical and cylindrical coordinate systems. Test every combination of _gradient_, _divergence_, _curl_ on every possible expression, verify that they raise exceptions when they should.
* Update the documentation.
* Make sure there is a `@deprecated` function.
* rename `differential_class` to `metric_lambda` and invert its elements.

> Instead of defining a differential_class we should define a scalar map to the general system via a set a functions:

There is a function which derives the metric from the transformation. See the changes. And... no, deriving the differential operators from the sets of transformations requires usage of equation solvers beyond the current capability of SymPy.

Comment (jbbskinny) [inline sympy/vector/functions.py]: > There is a function which derives the metric from the transformation.

Can you point me to it?  I can't seem to find it.

> And... no, deriving the differential operators from the sets of transformations requires usage of equation solvers beyond the current capability of SymPy.

Don't you feel that we should implement this before we continue?  For the purposes of scalability, I honestly feel this is the best route; not to mention the basic building blocks for all coordinate system transformations.  If we have a good curvilinear coordinate class the rest is cake, and I really don't see how speed is going to be an issue, at least in dim 3.  It's only a few partial derivatives and a square root to get the Lamé coefficients.

File: sympy/vector/coordsysrect.py [hunk 4]
```diff
@@ -1,19 +1,45 @@
...
+from sympy import eye, trigsimp, ImmutableMatrix as Matrix, Symbol, symbols, sqrt, Lambda, Dummy
 from sympy.core.compatibility import string_types, range
 from sympy.core.cache import cacheit
 from sympy.vector.orienters import (Orienter, AxisOrienter, BodyOrienter,
                                     SpaceOrienter, QuaternionOrienter)
-import sympy.vector
 
 
-class CoordSysCartesian(Basic):
+def build_instance_from_lambda(fun=lambda x, y, z: (x, y, z)):
```
Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: This is the function to calculate the Lamé parameters @jbbskinny , though they are returned in an unusual way.

I chose to return a _Lambda_, because it does not require to define symbols. An alternative would have been to return a triple of symbols, say _r, theta, phi_ and their expression, say `1, r**2, r**2*sin(theta)**2`.

Try it yourself: `build_instance_from_lambda(lambda r, theta, phi: (r*sin(phi)*cos(theta), r*sin(phi)*sin(theta), r*cos(phi)))`

Comment (jbbskinny) [inline sympy/vector/coordsysrect.py]: OK.  The metric simplification does take a long time.  When I did it by hand in iSympy it didn't take near as long.  I'll get going on the changes for the metric_lambda, depreciation, tests.

Should we keep CoordSys3D() while the class handles only dim(3,) or go ahead and change the name to CoordSystem()?

Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: `CoordSystem` is already taken in `sympy.diffgeom`. Anyway, it's better to jot down some API proposal and ask @moorepants what he thinks. He's the leading developer of the vector module.

Comment (jbbskinny) [inline sympy/vector/coordsysrect.py]: @moorepants What should this proposal look like? Should I send via email?

File: sympy/vector/functions.py (hunk 3 above)
Comment (Upabjojr) [inline sympy/vector/functions.py]: > Don't you feel that we should implement this before we continue?

@jbbskinny try to play around with `sympy.diffgeom`, especially the way how coordinate systems are connected there:

http://docs.sympy.org/dev/modules/diffgeom.html#sympy.diffgeom.CoordSystem

Consider just one think: all parameters passed to the class constructor have to be set on class initialization. Once an object has been created, its fields should not be modified. This is an important rule of SymPy, which unfortunately may restrict API choices.

This is how `.connect_to( )` is implemented:
https://github.com/sympy/sympy/blob/b6c26ae535323ff041864b47d2fcac743a763fbe/sympy/diffgeom/diffgeom.py#L215

I wouldn't go that way, but it's still worth comparing. I think the API should be flexible, some users may only desire to input the Lamé parameters, others may wish to specify the full transformation from another coordinate system. I believe we need an API which allows to do both, and I would play with the constructor arguments.

By the way... do you know there is a Wiki of SymPy? I'd suggest to draft some API change proposals before starting coding:
https://github.com/sympy/sympy/wiki

File: sympy/vector/coordsysrect.py (hunk 4 above)
Comment (Upabjojr) [inline sympy/vector/coordsysrect.py]: > depreciation

Are you sure that's the right spelling?
http://english.stackexchange.com/questions/45295/why-is-there-confusion-between-depreciated-and-deprecated

When I hear _depreciation_, I think of a lowering in prices, not of the process of rendering things obsolete.

Comment (jbbskinny) [inline sympy/vector/coordsysrect.py]: No, I'm using the ```@SymPyDeprecationWarning```.  I believe that is what @moorepants  suggested.  I still have to figure out how to document it correctly for PR.  

I'll work on the proposal now.  

It should surely be well rounded.  I have checked out the ```sympy.diffgeom``` library, but our coordinate systems will be globally valid, which comes back to the smooth/invertible transformations.  If the user is worried about regions of validity, they will be told to consult the ```sympy.diffgeom``` in our docs.

**Comment by Upabjojr** [comment]
OK, let's stop commenting on the code.

My PR adds a fuzzy way to express the reciprocals of Lamé parameters, through a _Lambda_ instance, as a parameter called _differential\_class_.

What about two parameters:
* _lame\_parameters_ which should act as my proposal, maybe not the reciprocals.
* _transformation\_from_ which defines the transformation from another coord system.

Both should be defaulted to `None`, if both are given, there should be an assertion that the Lamé parameters generated by the transformation are the same. What about this?

**Comment by jbbskinny** [comment]
I like it!  It is easy and flexible for the user.  I'll start coding when the proposal is submitted.
> lame_parameters which should act as my proposal, maybe not the reciprocals.

This will be great for speed!

> transformation_from which defines the transformation from another coord system.

I think this is a really good idea. It should also have a check if parent is defined, that way the origin and orientation is set for proper handling.  This will allow cylindrical -> spherical type transformations, using the curr system vars to make lambda function with transformation_from tuple.  If no parent is provided it will assume a cartesian -> whatever transformation.  Also, this is good for differential operators, seeing as how it provides linkage between coordinate systems.  Only thing I am worried about is the speed.. as you previously mentioned.

I'll send you my proposal as well.  I will probably just email it out.

**Comment by jbbskinny** [comment]
@moorepants Proposal:  

We will add the *transformation_to* and *lame_parameters* keyword arguments and deprecate CoordSysCartesian().  This keeps coordinate systems as one mathematical object and affords linkage between systems for differential operations.  Also, easy checking of un-linked coordinate systems when performing operations like ```gradient(x*r)```.

**Comment by moorepants** [comment]
> @moorepants What should this proposal look like? Should I send via email?

I don't really need a proposal for improving this package. You all are welcome to improve it. My only major concern is that you would change it such that we can never use it to replace sympy.physics.vector. Our goal with this package was to make a more general version of sympy.physics.vector based on immutable types. The goal was to have very general vectors which have a stronger mathematical base, but retain the ability to compute positions, velocities, accelerations, etc. from vectors that are built from vector components expressed in many different related reference frames.

My suggestion would be to learn about sympy.physics.vector and its capabilities so that you can keep them in mind when moving forward. Ultimately, we'd like to deprecate sympy.physics.vector in place of a more general, more efficient, faster, package that can also be used for a broader set of problems. For example, we hoped to be able to solve electrodynamics problems and flexible body systems eventually.

Here are some links that could get you started on the history. Check out Sachin and Prasoon's work:

https://github.com/sympy/sympy/wiki/GSoC-2014-Report
https://github.com/sympy/sympy/wiki/GSoC-2013-Report

**Comment by jbbskinny** [comment]
OK.  I will give those a deeper look-over.

The coordinate system class I am trying to push will be immutable in the sense that one has to define a new instance to change any properties of the coordinate system defined on the, as of now, three dimensional Euclidean space.  However, they can be linked by self.parent.  This new class will change none of the old methods (```self.scalar_map()``` not withstanding), only add support for a general curvilinear coordinate systems in a straightforward way.  If speed becomes a problem I have some experience in piping slow running Python processes to C applications.  Though I believe the slowest part is calculating the simplified metric ( not ```J.T * old_metric * J``` itself).  I plan to hard code lame_parameters for popular systems to eradicate the slow part.  Systems for problems with obscure BC's could take a bit longer to initialize.

**Comment by moorepants** [comment]
> I am worried about how general you want this vector module to be. If we want the most general form of a vector, a VectorSpace() will need to be initialized and then vectors can be defined on that space independent of any coordinate systems.

If it replicated what sympy.physics.vector does with immutable base types, then that's general enough for me. It's up to you how much more general you'd like to make things. Right now we only support 3D Cartesian coordinate systems, which is the same as sympy.physics.vector. But we did generalize here by adding in the variable unit scalars (not in sympy.physics.vector) so that vector spaces are supported. That and that we use immutable base types are the fundamental differences in sympy.vector and sympy.physics.vector at the moment.

**Comment by jbbskinny** [comment]
Well, for now I'll keep it in the realm of the real, and focus on complex vector spaces later.  A curvilinear coordinate system representation of dim(3) Euclidean space is a good, atomic PR.

**Comment by jbbskinny** [comment]
How does SymPy append to *self.args*?  I'm trying to add a str attribute related to the system type and free_symbols() throwing an AttributeError.

**Comment by Upabjojr** [comment]
You don't appends. SymPy objects must be immutable. _self.args_ are the arguments passed to the constructor of _Basic_, which is a superclass of all SymPy objects.

By the way, all _sympy.args_ arguments have to be instances of subclasses of Basic.

Write the argument parser/transformations in the constructor and the pass them to the super class constructor.

Make also sure that _a.func(*a.args) == a_ for every instance _a_ of your SymPy objects.

**Comment by jbbskinny** [comment]
Well, I'm curious as to why ```obj._name = name``` doesn't end up in *self.args* but ```obj._transformation_to = 'spherical'``` does.  Is it because it is a keyword arg?

Note: I just hardcoded for understanding. It is not that way in the code.

**Comment by Upabjojr** [comment]
> Well, I'm curious as to why obj._name = name doesn't end up in self.args but obj._transformation_to = 'spherical' does. Is it because it is a keyword arg?

That was the work of a previous author. The == operator matches the _.args_ in SymPy, he wanted the name to be excluded, so that == returns True if two coordinate systems have the same properties. That's an unfortunate choice, because == in SymPy should stand for structural equality.

https://github.com/sympy/sympy/wiki/Faq#why-does-sympy-say-that-two-equal-expressions-are-unequal

http://docs.sympy.org/latest/tutorial/gotchas.html#equals-signs

Anyway, correcting it would cause an API break, I would avoid that for now.

**Comment by Upabjojr** [comment]
Starting from tomorrow I will probably have scarce internet access for about a week. I may not answer in that period.

**Comment by jbbskinny** [comment]
I guess the only way around this then is to invoke a dictionary relating the keywords like ```'spherical``` to their relations lambda.  This will keep the speed and immutability of ```CoordSystem3D()```, only muck up ```__new__()```.

I was considering adding a metric property.  Would the ```@cacheit``` decorator help for multiple calls?  I would have it return the immutable matrix.  Just curious as to best practices for that decorator is all.

**Comment by Upabjojr** [comment]
Write code. I'll comment the code.

**Comment by jbbskinny** [comment]
@moorepants I see what you meant now about you being worried I'd change the code to much.  I'll be closing this PR (because of the long list of comments) and opening a new one.  Unfortunately, for the time being I'll only be able to change the lambda instance in @Upabjojr's PR.  That is unless I have a epiphany of sorts.

The thing is is I believe I can make it work with hacks but it'll really muck up the code, if you all are comfortable with that let me know but I prefer fluent code.  Expect the PR to be opened tomorrow, maybe the next day.

**Comment by Upabjojr** [comment]
> Unfortunately, for the time being I'll only be able to change the lambda instance in @Upabjojr's PR. That is unless I have a epiphany of sorts.

_Lambda_ is a good choice to pass in _.args_. I would simply add other possible constructing expressions, that may be transformed into a _Lambda_ in the constructor.

**Comment by jbbskinny** [comment]
The issue there is in ```locate_new()``` and ```orient_new()``` because it would use the Lambda instance in the return statement with no recollection of the keyword used (which is problematic for speed).  If I pass ```self._coord_sys_type```, where I saved the name of the coordinate system, it won't work for non-built-in coordinate systems (because only those name would be kept in the constructor dictionary).  So I would have to be consistent passing in only Lambda's or only built-in coordinate systems (where the constructor finds the Lamé lambda based off the constructor expression).

The only work around would be to have two dictionaries that are mirror images of each other, which would feel weird... unless there is a reverse hashing function I don't know about.

You'll see what I mean when I make the new PR.  I am going to clean up my current code and push it for early commenting.  Hopefully, you all see something that I don't.
